src/ty/ty_check/mod.rs

```diff
@@ -77,7 +77,11 @@ impl<'a> Context<'a> {
     /// impossible to satisfy.
     pub fn add_constraint(&mut self, expected: Ty, found: Ty, loc: Location) {
         let constr = Constraint::new(expected, found);
-        self.constraints.push(Located::new(constr, loc))
+        // New constraints are added at the begining because the `Unifier` starts processing
+        // constraints from last to first. If we just push the constraint, we end up taking care of
+        // the newer constraints first. Which are more complex and can end up in less readable type
+        // errors.
+        self.constraints.insert(0, Located::new(constr, loc))
```

Comment:
In this case wouldn't it be better to use a `VecDeque`[[1](https://doc.rust-lang.org/std/collections/struct.VecDeque.html)] and add new constraints to the back and remove them from the front?

Reply:
Oh yes, that would actually play well with the future incremental typing because we'd need to do sandwiched push and pops. Thanks